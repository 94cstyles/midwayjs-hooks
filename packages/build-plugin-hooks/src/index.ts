import { IPlugin } from '@alib/build-scripts'

const buildHooksRequest: IPlugin = (pluginApi) => {
  pluginApi.onGetWebpackConfig((config) => {
    const MidwayHooksLoader = require.resolve('@midwayjs/hooks-loader')
    ;['jsx', 'tsx'].forEach((type) => {
      config.module.rule(type).use('midway-hooks').loader(MidwayHooksLoader)
    })
  })
}

export default buildHooksRequest
