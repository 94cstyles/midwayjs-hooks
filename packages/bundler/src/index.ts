import omit from 'lodash/omit'
import { createUnplugin, UnpluginOptions } from 'unplugin'
import {
  AbstractRouter,
  ApiRoute,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  parseApiModule,
} from '@midwayjs/hooks-core'
import jiti from 'jiti'

jiti().register()

interface BundlerConfig {
  name: string
  router: AbstractRouter
}

export abstract class AbstractBundlerAdapter {
  constructor(private config: BundlerConfig) {}

  getName() {
    return this.config.name
  }

  getRouter() {
    return this.config.router
  }

  getUnplugin(): Partial<UnpluginOptions> {
    return {}
  }

  transformApiRoutes(apis: ApiRoute[]) {
    return apis
  }

  generateClient(apis: ApiRoute[]) {
    let fetcherId = 0
    const importClient = new Map<string, [string, string, number]>()

    const importCodes = new Set<string>()
    const functionCodes = []

    for (const {
      trigger,
      functionName,
      functionId,
      useInputMetadata,
    } of apis) {
      if (!trigger.requestClient) continue
      const { fetcher, client } = trigger.requestClient

      const clientId = `${client}_${fetcher}`
      if (!importClient.has(clientId)) {
        importClient.set(clientId, [client, fetcher, fetcherId++])
      }

      const [clientName, fetcherName, idx] = importClient.get(clientId)
      const uniqueFetcher = `rpc$${idx}`
      importCodes.add(
        `import { ${fetcherName} as ${uniqueFetcher} } from '${clientName}'`
      )

      const exportModifier =
        functionName === EXPORT_DEFAULT_FUNCTION_ALIAS ? 'default' : ''
      functionCodes.push(`
        export ${exportModifier} function ${functionName} (...args) {
          const route = ${functionName}.route;
          return ${uniqueFetcher}(...args, route)
        };
        ${functionName}.route = {
          trigger: ${JSON.stringify(omit(trigger, 'requestClient'))},
          functionId: '${functionId}',
          useInputMetadata: ${!!useInputMetadata},
        };
      `)
    }

    const code = [
      '// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.',
      ...importCodes,
      ...functionCodes,
    ].join('\n')
    return this.formatCode(code)
  }

  private formatCode(code: string) {
    try {
      const prettier = require('prettier')
      return prettier.format(code, {
        semi: true,
        singleQuote: true,
        parser: 'babel',
      })
    } catch (e) {
      console.log('format code error', e)
      return code
    }
  }
}

export function createBundlerPlugin(adapter: AbstractBundlerAdapter) {
  return createUnplugin((options) => {
    return {
      name: adapter.getName(),
      enforce: 'pre',
      transformInclude(id: string) {
        return adapter.getRouter().isApiFile(id)
      },
      async transform(code: string, id: string) {
        const apis = parseApiModule(require(id), id, adapter.getRouter())
        const transformedApis = adapter.transformApiRoutes(apis)
        return adapter.generateClient(transformedApis)
      },
      ...adapter.getUnplugin(),
    }
  })
}
