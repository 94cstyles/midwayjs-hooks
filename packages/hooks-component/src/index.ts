import { createConfiguration, IMidwayContainer } from '@midwayjs/core'

import { Inject, Controller, Get, Post, Provide } from '@midwayjs/decorator'

import { als } from '@midwayjs/hooks/lib/api/async_hooks/als'
import { EnhancedFunc } from '@midwayjs/hooks/lib/hoc/type'
import {
  WebRouterConfig,
  WebRouter,
  getFunctionId,
} from '@midwayjs/hooks-router'

interface HooksConfig extends Omit<WebRouterConfig, 'source'> {}

export const fnMap = new Map<EnhancedFunc, any>()

export const createHooks = (config: HooksConfig) => {
  const configuration = createConfiguration({
    namespace: 'hooks',
    directoryResolveFilter: createResolveFilter(config),
  })
    .onReady(() => {})
    .onStop(() => {})

  return {
    Configuration: configuration,
  }
}

function createResolveFilter(config: HooksConfig) {
  return config.routes.map((route) => {
    return {
      pattern: route.baseDir,
      filter: (
        mod: {
          default?: any
          [key: string]: any
        },
        sourceFilePath: string,
        container: IMidwayContainer
      ) => {
        const root: string = container.get('appDir')
        const router = new WebRouter(root, {
          routes: config.routes,
          source: (config as any).source,
        })

        // export default
        const defaultExports = mod.default || mod
        if (typeof defaultExports === 'function') {
          createFunctionContainer({
            container,
            router,
            fn: defaultExports,
            sourceFilePath,
            isExportDefault: true,
          })
        }

        for (const key of Object.keys(mod)) {
          if (key === 'default') {
            continue
          }

          const value = mod[key]
          if (typeof value === 'function') {
            createFunctionContainer({
              container,
              router,
              fn: value,
              sourceFilePath,
              isExportDefault: false,
            })
          }
        }
      },
    }
  })
}

function createFunctionContainer(config: {
  container: IMidwayContainer
  fn: EnhancedFunc
  sourceFilePath: string
  router: WebRouter
  isExportDefault: boolean
}) {
  const { container, fn, router, sourceFilePath, isExportDefault } = config

  const fnName = isExportDefault ? '$default' : fn.name
  const id = getFunctionId({
    router,
    sourceFilePath,
    functionName: fnName,
    isExportDefault,
  })

  const containerId = 'hooks_func::' + id
  const httpPath = router.getHTTPPath(sourceFilePath, fnName, isExportDefault)
  const HttpMethod = fn.length === 0 ? Get : Post

  fnMap.set(fn, {
    containerId,
    httpPath,
    httpMethod: fn.length === 0 ? 'GET' : 'POST',
  })

  @Provide(containerId)
  @Controller('/')
  class FunctionContainer {
    @Inject()
    ctx: any

    @HttpMethod(httpPath, { middleware: fn.middleware || [] })
    async handler() {
      const bindCtx = {
        ctx: this.ctx,
      }

      let args = this.ctx?.request?.body?.args || []
      if (typeof args === 'string') {
        args = JSON.parse(args)
      }

      return await als.run(bindCtx, () => fn(...args))
    }
  }

  container.bind(containerId, FunctionContainer)
}
