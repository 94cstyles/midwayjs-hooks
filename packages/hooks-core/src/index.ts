export * from './router'
export * from './component'
export * from './runtime'
export * from './request'
export * from './request/generate'
export * from './config'
export * from './proxy'

export * from './types/common'
export * from './types/http'
export * from './types/serverless'
export * from './types/config'
