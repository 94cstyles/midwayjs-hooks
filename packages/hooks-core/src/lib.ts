import superjson from 'superjson'
import consola from 'consola'

export { superjson, consola }
