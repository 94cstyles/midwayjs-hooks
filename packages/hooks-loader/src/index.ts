import { LambdaHTTPMethod } from '@midwayjs/hooks-shared'
import { codeFrameColumns } from '@babel/code-frame'
import * as parser from '@babel/parser'
import traverse, { NodePath } from '@babel/traverse'
import * as t from '@babel/types'
import { Identifier, FunctionDeclaration } from '@babel/types'
import { loader } from 'webpack'
import { buildRequest, RenderParam } from './render'
import { helper, getFunctionHandlerName, BuiltinEnhancer } from '@midwayjs/next-hooks-compiler'
import { debug } from './util'

export default async function loader(this: loader.LoaderContext, source: string) {
  const callback = this.async()
  const resourcePath = this.resourcePath

  if (resourcePath.includes('node_modules')) {
    return callback(null, source)
  }

  const root = this.rootContext || (this as any).options?.context
  helper.root = root

  if (!helper.isLambdaFile(resourcePath)) {
    return callback(null, source)
  }

  debug('compile %s', resourcePath)

  let ast: t.File
  try {
    ast = parser.parse(source, {
      sourceType: 'module',
      plugins: [
        'typescript',
        'classProperties',
        'objectRestSpread',
        'optionalCatchBinding',
        'dynamicImport',
        'decorators-legacy',
        'asyncGenerators',
        'exportDefaultFrom',
        'exportNamespaceFrom',
        'optionalCatchBinding',
        'throwExpressions',
        'optionalChaining',
        'nullishCoalescingOperator',
      ],
    })
  } catch (error) {
    error.message = codeFrameColumns(source, error.loc, {
      highlightCode: true,
      message: error.message,
    })
    throw error
  }

  const funcs: RenderParam[] = []

  let err: Error

  /**
   * Cases in which export is allowed
   *
   * 1. export async function demo () {}
   * 2. export const demo = async function () {}
   * 3. export const demo = async () => {}
   * 4. export default async function demo () {}
   * 5. export default async function () {}
   * 6. export default async () => {}
   */
  traverse(ast, {
    ExportNamedDeclaration(path) {
      const func: RenderParam = {}
      const declaration = path.node.declaration

      if (t.isTSTypeAliasDeclaration(declaration) || t.isTSInterfaceDeclaration(declaration)) {
        return
      }

      if (t.isFunctionDeclaration(declaration)) {
        const id = declaration.id

        func.url = helper.getHTTPPath(resourcePath, id.name, false)

        const { method, params } = parseFunctionParams(declaration.params)
        func.method = method

        func.meta = {
          functionName: getFunctionHandlerName({
            sourceFilePath: resourcePath,
            isExportDefault: false,
            functionName: id.name,
          }),
          unstable_params: params,
        }
        func.functionId = id.name

        funcs.push(func)
        return
      }

      if (t.isVariableDeclaration(declaration)) {
        for (const variableDeclarator of declaration.declarations) {
          const id = variableDeclarator.id as Identifier
          const init = variableDeclarator.init

          func.url = helper.getHTTPPath(resourcePath, id.name, false)
          func.meta = {
            functionName: getFunctionHandlerName({
              sourceFilePath: resourcePath,
              isExportDefault: false,
              functionName: id.name,
            }),
          }
          func.functionId = id.name

          if (t.isArrowFunctionExpression(init) || t.isFunctionExpression(init)) {
            const { method, params } = parseFunctionParams(init.params)
            func.method = method
            func.meta.unstable_params = params
          } else if (t.isCallExpression(init) && BuiltinEnhancer.includes((init.callee as Identifier).name)) {
            const callback = init.arguments[1] as t.FunctionExpression
            const { method, params } = parseFunctionParams(callback.params)
            func.method = method
            func.meta.unstable_params = params
          } else {
            err = buildCodeFrameError(path, getErrorMessage())
            path.stop()
            return
          }

          funcs.push(func)
        }
        return
      }

      err = buildCodeFrameError(path, getErrorMessage())
      path.stop()
    },
    ExportDefaultDeclaration(path) {
      const declaration = path.node.declaration
      const func: RenderParam = {}

      let funcExpression: t.FunctionDeclaration | t.FunctionExpression | t.ArrowFunctionExpression = null

      if (t.isFunctionDeclaration(declaration) || t.isArrowFunctionExpression(declaration)) {
        funcExpression = declaration
      } else if (t.isCallExpression(declaration) && BuiltinEnhancer.includes((declaration.callee as Identifier).name)) {
        funcExpression = declaration.arguments[1] as t.FunctionExpression
      }

      if (!funcExpression) {
        err = buildCodeFrameError(path, getErrorMessage())
        path.stop()
        return
      }

      const functionName = getFunctionHandlerName({
        sourceFilePath: resourcePath,
        isExportDefault: true,
        functionName: getFunctionName(path),
      })

      const { method, params } = parseFunctionParams(funcExpression.params)
      func.method = method
      func.functionId = getFunctionName(path)
      func.isExportDefault = true
      func.url = helper.getHTTPPath(resourcePath, functionName, true)
      func.meta = {
        functionName,
        unstable_params: params,
      }

      funcs.push(func)
    },
  })

  if (err) {
    callback(err)
    return
  }

  const code = buildRequest(funcs, root, this.query)
  debug('compile %s', resourcePath)
  callback(null, code)
}

function getFunctionName(path: NodePath<t.ExportDefaultDeclaration>) {
  const declaration = path.node.declaration
  const defaultExportName = '$default'

  if (t.isFunctionDeclaration(declaration)) {
    return declaration?.id?.name || defaultExportName
  }

  return defaultExportName
}

function parseFunctionParams(params: FunctionDeclaration['params']) {
  const keys = []

  for (const param of params) {
    // (a) => {}
    if (t.isIdentifier(param)) {
      keys.push(param.name)
      continue
    }

    //  (a = 1) => {}
    if (t.isAssignmentPattern(param) && t.isIdentifier(param.left)) {
      keys.push(param.left.name)
      continue
    }

    // (...a) => {}
    if (t.isRestElement(param) && t.isIdentifier(param.argument)) {
      keys.push(param.argument.name)
      continue
    }

    console.log('Function parameter parsing error, node type: %s', param.type)
  }

  return {
    method: params?.length > 0 ? 'POST' : ('GET' as LambdaHTTPMethod),
    params: keys,
  }
}

function getErrorMessage() {
  return `
    When a function file is used as an API, only the following export methods are allowed
      1. export async function demo () {}
      2. export const demo = async function () {}
      3. export const demo = async () => {}
      4. export default async function demo () {}
      5. export default async function () {}
      6. export default async () => {}
  `
}

function buildCodeFrameError(path: NodePath<any>, message: string) {
  const root = path.findParent((p) => !p.parentPath)
  const frame = codeFrameColumns(
    root.toString(),
    {
      end: path.node.loc.end,
    },
    {
      message: message,
    }
  )

  return new Error(frame)
}
