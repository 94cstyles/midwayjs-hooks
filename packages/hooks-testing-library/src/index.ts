import { createApp as createWebApp, close } from '@midwayjs/mock'
import { IMidwayApplication } from '@midwayjs/core'
import {
  getConfig,
  getProjectRoot,
  als,
  HooksContext,
} from '@midwayjs/hooks-core'
import { join } from 'path'
import { remove } from 'fs-extra'

export async function createApp(baseDir?: string) {
  const root = getProjectRoot(baseDir)
  const config = getConfig(baseDir)
  // TODO createFunctionApp
  const app: IMidwayApplication<any> = await createWebApp(
    root,
    {},
    config.framework
  )
  return new HooksApplication(app)
}

class HooksApplication {
  private app: IMidwayApplication<any>
  constructor(app: IMidwayApplication<any>) {
    this.app = app
  }

  // TODO Allow pass user define context
  // TODO Should invoke function call
  async runFunction<T extends (...args: any) => any>(func: T) {
    return await als.run(this.mockContext(), () => func())
  }

  async runHooks<T extends (...args: any) => any>(hooks: T) {
    return await als.run(this.mockContext(), () => hooks())
  }

  request() {}

  mockContext(): HooksContext {
    return {
      ctx: {},
    }
  }

  async close() {
    await close(this.app)
    if (process.env.EGG_HOME) {
      await remove(join(process.env.EGG_HOME, 'logs'))
    }
    const appDir = this.app?.getAppDir()
    if (appDir) {
      await remove(join(appDir, 'logs'))
      await remove(join(appDir, 'run'))
    }
  }
}
