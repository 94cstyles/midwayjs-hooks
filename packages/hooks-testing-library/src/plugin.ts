import { SuperAgentRequest } from 'superagent'
import { superjson } from '@midwayjs/hooks-core'

export function SuperJSONPlugin() {
  return (req: SuperAgentRequest) => {
    req.on('response', (res) => {
      if (res.body) res.body = superjson.deserialize(res.body)
      if (res.text) res.text = superjson.parse(res.text)
    })
  }
}
