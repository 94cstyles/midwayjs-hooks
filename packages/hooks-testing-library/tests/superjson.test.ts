import path from 'path'

import { createApp, HooksApplication } from '../src'
import { CustomError } from './fixtures/superjson/src/error/custom'
import {
  returnString,
  returnObject,
  getContext,
  get,
  post,
  returnError,
  throwError,
  throwCustomError,
} from './fixtures/superjson/src/lambda/index'

let app: HooksApplication
beforeAll(async () => {
  app = await createApp({
    root: path.join(__dirname, './fixtures/superjson'),
  })
})

afterAll(async () => {
  await app.close()
})

describe('Hooks Api', () => {
  test('HTTP GET', async () => {
    expect(await app.runFunction(get)).toMatchSnapshot()
  })

  test('HTTP POST', async () => {
    expect(await app.runFunction(post, 'Lxxyx')).toMatchSnapshot()
  })

  test('return string', async () => {
    expect(await app.runFunction(returnString)).toMatchSnapshot()
  })

  test('return object', async () => {
    expect(await app.runFunction(returnObject)).toMatchSnapshot()
  })

  test('get request context', async () => {
    expect(await app.runFunction(getContext)).toMatchSnapshot()
  })

  test('return error', async () => {
    expect(await app.runFunction(returnError)).toMatchSnapshot()
  })

  test('throw error', async () => {
    try {
      await app.runFunction(throwError)
    } catch (error) {
      expect(error instanceof Error)
      expect(error).toMatchSnapshot()
    }
  })

  test('throw custom error', async () => {
    try {
      await app.runFunction(throwCustomError)
    } catch (error) {
      expect(error instanceof CustomError)
      expect(error).toMatchSnapshot()
    }
  })
})
