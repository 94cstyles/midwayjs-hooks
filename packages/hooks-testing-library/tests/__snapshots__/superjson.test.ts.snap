// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks Api HTTP GET 1`] = `"GET"`;

exports[`Hooks Api HTTP POST 1`] = `
Object {
  "method": "POST",
  "name": "Lxxyx",
}
`;

exports[`Hooks Api get request context 1`] = `
Object {
  "path": "/api/getContext",
}
`;

exports[`Hooks Api return error 1`] = `
Object {
  "err": [Error: this is error],
}
`;

exports[`Hooks Api return object 1`] = `
Object {
  "message": "Hello World!",
}
`;

exports[`Hooks Api return string 1`] = `"Hello World"`;

exports[`Hooks Api throw custom error 1`] = `[Error]`;

exports[`Hooks Api throw error 1`] = `[Error: this is error]`;
